src/lib/ketrew_configuration.ml

```diff
@@ -165,98 +162,156 @@ let server
+let database_parameters e = e.database_parameters
+let persistent_state_key e = e.persistent_state_key
+let is_unix_ssh_failure_fatal e = e.turn_unix_ssh_failure_into_target_failure
+let mode t = t.mode
+let standalone_engine st = st.standalone_engine
+let server_engine s = s.server_engine
+let connection c = c.connection
+let token c = c.token
+
+let standalone_of_server s =
+  {standalone_ui = s.server_ui;
+   standalone_engine = s.server_engine;}
+
+module File = struct
+  type configuration = t
+   [@@deriving yojson]
+  type profile = {
+    name: string;
+    configuration: configuration;
+  } [@@deriving yojson]
+  type t = [
+    | `Ketrew_configuration [@name "Ketrew"] of profile list
+  ] [@@deriving yojson]
+
+  let parse_string_exn s =
+    match Yojson.Safe.from_string s |> of_yojson with
+    | `Ok o -> o
+    | `Error e -> failwith (fmt "Configuration parsing error: %s" e)
+
+  let to_string s = to_yojson s |> Yojson.Safe.pretty_to_string ~std:true
+
+  let get_profile t the_name =
+    match t with
+    | `Ketrew_configuration profiles ->
+      List.find_map profiles ~f:(fun {name; configuration} ->
+          if name = the_name then Some configuration else None)
 
+  let pick_profile_exn ?name t =
+    let name =
+      match name with
+      | Some n -> n
+      | None ->
+        try Sys.getenv "KETREW_PROFILE" with
+        | _ -> "default"
+    in
+    get_profile t name
+    |> Option.value_exn ~msg:(fmt "profile %S not found" name)
+
+  let default_ketrew_path =
+    Sys.getenv "HOME" ^ "/.ketrew/"
+
+  let default_configuration_filenames = [
+    "configuration.json";
+    "configuration.ml";
+    "configuration.sh";
+    "configuration.url";
+  ]
+
+  let get_path ?root () =
+    let env n () = try Some (Sys.getenv n) with | _ -> None in
+    let try_options l last_resort =
+      match List.find_map l ~f:(fun f -> f ()) with
+      | Some s -> s
+      | None -> last_resort () in
+    let findout_path () =
+      try_options [
+        (fun () -> root);
+        env "KETREW_ROOT";
+      ] (fun () -> default_ketrew_path) in
+    let find_in_path () =
+      let ketrew_path = findout_path () in
+      try_options
+        (List.map default_configuration_filenames
+           ~f:(fun path ->
+               fun () -> if Sys.file_exists path then Some path else None))
+        (fun () -> ketrew_path ^ "configuration.json")
+    in
+    try_options [
+      env "KETREW_CONFIGURATION";
+      env "KETREW_CONFIG";
+    ]
+      find_in_path
+
+  let read_file_no_lwt path =
+    let i = open_in path in
+    begin try
+      let content =
+        let buf = Buffer.create 1023 in
+        let rec get_all () =
+          begin try
+            let line = input_line i in
+            Buffer.add_string buf (line ^ "\n");
+            get_all ()
+          with e -> ()
+          end;
+        in
+        get_all ();
+        Buffer.contents buf in
+      close_in i;
+      content
+    with e -> close_in i; raise e
```

Comment:
What exceptions can would escape the inner ones? Seems like you're guarding for `Buffer.contents buf` ?